src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs

```diff
@@ -747,6 +747,13 @@ public override TypeProvider FindTypeProviderForSchema(Schema schema)
         public override CSharpType? FindTypeByName(string originalName)
         {
             _schemaOrNameToModels.TryGetValue(originalName, out TypeProvider? provider);
+
+            if (provider == null)
+            {
+                // Try to search declaration name too if no key matches. i.e. Resource Data Type will be appended a 'Data' in the name and won't be found through key
+                provider = _schemaOrNameToModels.FirstOrDefault(s => s.Value is MgmtObjectType mot && mot.Declaration.Name == originalName).Value;
```

Comment:
I did not get why we need this - this scenario "Resource Data Type will be appended a 'Data' in the name and won't be found through key" is handled on line 750 (before change)

Reply:
You see two dictionaries here: `_schemaOrNameToModels` and `ResourceSchemaMap`
`ResourceSchemaMap` is a sub-dictionary from `AllSchemaMap`.
And `AllSchemaMap` is initialized [here](https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs#L116) by this `InitializeModels` method, and this method is returning the `_schemaOrNameToModels` dictionary (there is an implicit operator taking effect during the return).
Therefore looking things up inside `_schemaOrNameToModels` is the same as looking things up in `AllSchemaMap`.

This is the reason why I wonder why we have to do this, because it is happening below.

But if there is another case of this happening, when the schema.Name is different from the name of the corresponding type provider, we should remove line 750. Now the only case of this is "ResourceData".